File: 1094. Car Pooling.java
// Medium
// Difference Array
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/car-pooling/

class Solution {
    public boolean carPooling(int[][] trips, int capacity) {
        int[] diff = new int[1001];

        for (int[] t : trips) {
            int num = t[0], from = t[1], to = t[2];
            diff[from] += num;
            diff[to] -= num;
        }

        int cur = 0;
        for (int i = 0; i < 1001; i++) {
            cur += diff[i];
            if (cur > capacity)
                return false;
        }
        return true;
    }
}

File: 1001-2000/1094. Car Pooling.py
# Time:O(n+m), Space:O(m)

from typing import List

class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        max_loc = 0
        for num, start, end in trips:
            max_loc = max(max_loc, end)

        diff = [0] * (max_loc + 1)

        for num, start, end in trips:
            diff[start] += num
            diff[end] -= num

        cur = 0
        for change in diff:
            cur += change
            if cur > capacity:
                return False

        return True
    